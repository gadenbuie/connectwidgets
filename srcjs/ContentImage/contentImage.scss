.rscimage {
  align-items: center;
  background: white;
  display: flex;
  flex: 1;
  justify-content: center;
  position: relative;
  width: 100%;

  &__img {
    background-size: cover;
    background-position: center center;
    background-repeat: no-repeat;
    margin: 0;
    padding: 36% 0;
    width: 100%;
  }

  &__request-access {
    align-items: center;
    background: #EDF6FF;
    border-radius: 1.24em;
    box-shadow: 0 0 3px 1px rgba(0,0,0, 0.2);
    color: #5882B2;
    display: none;
    font-size: 0.8em;
    font-weight: 700;
    height: 2.48em;
    line-height: 2.48em;
    padding: 0 1.2em;
    position: absolute;

    &::before {
      background-image: url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNCAyNCI+PHBhdGggZmlsbD0iIzU4ODJCMiIgZD0iTTEyLDEzYTEuNDksMS40OSwwLDAsMC0xLDIuNjFWMTdhMSwxLDAsMCwwLDIsMFYxNS42MUExLjQ5LDEuNDksMCwwLDAsMTIsMTNabTUtNEg5VjdhMywzLDAsMCwxLDUuMTItMi4xMywzLjA4LDMuMDgsMCwwLDEsLjc4LDEuMzgsMSwxLDAsMSwwLDEuOTQtLjUsNS4wOSw1LjA5LDAsMCwwLTEuMzEtMi4yOUE1LDUsMCwwLDAsNyw3VjlhMywzLDAsMCwwLTMsM3Y3YTMsMywwLDAsMCwzLDNIMTdhMywzLDAsMCwwLDMtM1YxMkEzLDMsMCwwLDAsMTcsOVptMSwxMGExLDEsMCwwLDEtMSwxSDdhMSwxLDAsMCwxLTEtMVYxMmExLDEsMCwwLDEsMS0xSDE3YTEsMSwwLDAsMSwxLDFaIi8+PC9zdmc+");
      background-size: contain;
      background-position: center center;
      background-repeat: no-repeat;
      display: inline-block;
      content: "";
      height: 1.44em;
      margin-right: 0.4em;
      width: 1.44em;
    }
  }

  &--no-access::before {
    background: black;
    opacity: 0.7;
    content: "";
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  }

  &--no-access &__request-access {
    display: flex;
  }
}
