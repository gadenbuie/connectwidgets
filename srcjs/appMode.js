import ApiImg from '@/images/api.svg';
import AppImg from '@/images/app.svg';
import DocImg from '@/images/doc.svg';
import ModelImg from '@/images/model.svg';
import PinImg from '@/images/pin.svg';
import PlotImg from '@/images/plot.svg';

const appsModeMap = {
  // internal types to type name
  'api': 'API',
  'shiny': 'Application',
  'rmd-shiny': 'Document',
  'rmd-static': 'Document',
  'jupyter-static': 'Document',
  'static': {
    'plot': 'Plot',
    'pin': 'Pin',
  },
  'tensorflow-saved-model': 'Model',
  'python-api': 'API',
  'python-dash': 'Application',
  'python-streamlit': 'Application',
  'python-bokeh': 'Application',

  // name to applicable internal types
  'API': ['api', 'python-api'],
  'Document': ['rmd-shiny', 'rmd-static', 'jupyter-static'],
  'Application': ['shiny', 'python-dash', 'python-streamlit', 'python-bokeh'],
  'Plot': [['static', 'plot']],
  'Pin': [['static', 'pin']],
  'Model': ['tensorflow-saved-model'],
};

/**
 * Get human friendly content type name from internal type id.
 * @param {string} id Connect's internal content type id. E.g "jupyter-static".
 * @param {string} category Connect's internal sub-type content type id. E.g "pin".
 * @returns {string} A human friendly content type.
 */
export const appTypeOf = (id, category) => {
  let type = appsModeMap[id];
  if (category) {
    type = type[category];
  }
  // static type with no sub-category
  if (typeof type !== 'string') {
    type = 'Other';
  }
  return type || 'Other';
};

/**
 * Get the default image for a given content app mode.
 * @param {string} appMode The content app mode.
 * @param {string} contentCategory Connect's internal sub-type content type id. E.g "pin".
 * @returns The default image.
 */
export const appModeImg = (appMode, contentCategory) => {
  const type = appTypeOf(appMode, contentCategory);

  switch (type) {
    case 'API':
      return ApiImg;
    case 'Document':
      return DocImg;
    case 'Plot':
      return PlotImg;
    case 'Pin':
      return PinImg;
    case 'Model':
      return ModelImg;
    default:
      return AppImg;
  }
};
