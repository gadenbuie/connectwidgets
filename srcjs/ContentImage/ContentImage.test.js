import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { appModeImg } from '@/appMode';
import ContentImage from './ContentImage';

jest.mock('@/appMode');
appModeImg.mockReturnValue('default-img-url');

const imageUrl = 'https://science.rstudio.com/__api__/applications/7e4bdcb4-c809-40da-9717-be86625ee690/image';
const contentUrl = 'https://science.rstudio.com/content/7e4bdcb4-c809-40da-9717-be86625ee690';

const waitForCompleteLoad = async wrapper => {
  await Promise.resolve();
  if (!wrapper.instance().state.loaded) {
    await waitForCompleteLoad(wrapper);
  }
};

const mkWrapper = async () => {
  const wrapper = shallow(<ContentImage
    imageUrl={imageUrl}
    contentUrl={contentUrl}
    contentType='Document'
  />);
  await waitForCompleteLoad(wrapper);
  return wrapper;
};

describe('<ContentImage />', () => {

  it('renders properly', async () => {
    const wrapper = await mkWrapper();
    expect(fetch).toHaveBeenCalledWith(imageUrl);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.instance().state.imgUrl).not.toBe('default-img-url');
  });

  describe('forbidden images', () => {
    it('uses default image and request access label shows', async () => {
      fetch.mockResolvedValueOnce({ status: 403 });
      const wrapper = await mkWrapper();
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.instance().state.imgUrl).toBe('default-img-url');
      expect(wrapper.hasClass('rscimage--no-access')).toBe(true);
    });
  });

  describe('non existent images', () => {
    it('uses default image and request access label shows', async () => {
      fetch.mockResolvedValueOnce({ status: 204 });
      const wrapper = await mkWrapper();
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.instance().state.imgUrl).toBe('default-img-url');
      expect(wrapper.hasClass('rscimage--no-access')).toBe(false);
    });
  });
});
