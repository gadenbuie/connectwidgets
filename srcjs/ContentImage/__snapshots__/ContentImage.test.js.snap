// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentImage /> forbidden images uses default image and request access label shows 1`] = `
<a
  className="rscimage rscimage--no-access"
  href="https://science.rstudio.com/content/7e4bdcb4-c809-40da-9717-be86625ee690"
  rel="noreferrer"
  target="_blank"
>
  <figure
    className="rscimage__img"
    style={
      Object {
        "backgroundImage": "url(\\"default-img-url\\")",
      }
    }
  />
  <span
    className="rscimage__request-access"
  >
    REQUEST ACCESS
  </span>
</a>
`;

exports[`<ContentImage /> non existent images uses default image and request access label shows 1`] = `
<a
  className="rscimage"
  href="https://science.rstudio.com/content/7e4bdcb4-c809-40da-9717-be86625ee690"
  rel="noreferrer"
  target="_blank"
>
  <figure
    className="rscimage__img"
    style={
      Object {
        "backgroundImage": "url(\\"default-img-url\\")",
      }
    }
  />
  <span
    className="rscimage__request-access"
  >
    REQUEST ACCESS
  </span>
</a>
`;

exports[`<ContentImage /> renders properly 1`] = `
<a
  className="rscimage"
  href="https://science.rstudio.com/content/7e4bdcb4-c809-40da-9717-be86625ee690"
  rel="noreferrer"
  target="_blank"
>
  <figure
    className="rscimage__img"
    style={
      Object {
        "backgroundImage": "url(\\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNCAyNCI+PHBhdGggZmlsbD0iIzU4ODJCMiIgZD0iTTEyLDEzYTEuNDksMS40OSwwLDAsMC0xLDIuNjFWMTdhMSwxLDAsMCwwLDIsMFYxNS42MUExLjQ5LDEuNDksMCwwLDAsMTIsMTNabTUtNEg5VjdhMywzLDAsMCwxLDUuMTItMi4xMywzLjA4LDMuMDgsMCwwLDEsLjc4LDEuMzgsMSwxLDAsMSwwLDEuOTQtLjUsNS4wOSw1LjA5LDAsMCwwLTEuMzEtMi4yOUE1LDUsMCwwLDAsNyw3VjlhMywzLDAsMCwwLTMsM3Y3YTMsMywwLDAsMCwzLDNIMTdhMywzLDAsMCwwLDMtM1YxMkEzLDMsMCwwLDAsMTcsOVptMSwxMGExLDEsMCwwLDEtMSwxSDdhMSwxLDAsMCwxLTEtMVYxMmExLDEsMCwwLDEsMS0xSDE3YTEsMSwwLDAsMSwxLDFaIi8+PC9zdmc+\\")",
      }
    }
  />
  <span
    className="rscimage__request-access"
  >
    REQUEST ACCESS
  </span>
</a>
`;
